blog/_src/posts/2016-11-02-beta-reduction-part-1.md

```diff
@@ -0,0 +1,189 @@
+
+Now that we've discussed the syntax of the language, we can look at the semantics, or how terms evaluate.  Below I have the evaluation rules for the λ-calculus.  The arrow represents β-reduction which is the subject of this post.  The semantics rely on substitution which is depicted with brackets.  I have also defined the substitution function.  I'm assuming the Barendregt variable convention which states that every bound variable is distinct from every free variable.
+
+```
+x[ x := e ] = e
+x[ y := e ] = y
+(λx.e1)[ x := e2 ] = (λx.e1[ x := e2 ])
+(e1 e2)[ x := e3 ] = (e1[ x := e3 ] e2[ x := e3 ])
+```
+
+With the substitution function defined, we can write a semantics for evaluation:
+
+```
+------------------------------
+  (λx.e1) e2 ->β e1[ x := e2 ]
+
+    e1 ->β e1'
+--------------------
+  e1 e2 ->β e1' e2
+
+    e2 ->β e2
+--------------------
+  e1 e2 ->β e1 e2'
+```
+
+These rules mean that if you have a call in which you have a function applied to the argument, you substitute the argument into the function.  The next two rules say that if you can perform an evaluation step in either the first or second term, you can perform that step.  Notice that both the second and third rules might apply to a single term.  These rules are nondeterministic and in these cases it is fine to use whichever rule you want.
+
+# What is β-reduction?
+
+More generally, what is reduction?  Intuitively, a reduction system is a set of rules that determine how a term is stepped forwards in the computation.  β-reduction is Church's name for removing the λ from a function and substituting the argument for the parameter in the function body.  More formally, we can define β-reduction as follows:
+
+```
+(λx.e1) e2 = e1[ x := e2 ]
+```
+
+Given that definition, lets look at a simple example.  I've written the examples in the λ-calculus, Racket, OCaml, and Haskell.  It's important to note that these languages have more restricted semantics than the original λ-calculus.  In Racket and OCaml, it is not possible to substitute with anything except for a value, meaning you need to evaluate the argument until it is a function before substituting.  This makes the evaluation reduce left to right before substituting.  Haskell, in contrast, no reduction occurs in arguments, meaning that we would omit the third rule.  
+
```

Comment:
You should point out that call-by-need, which is how Haskell's reduction is usually implemented, cannot be described directly in this way: it needs more syntax (let-bindings to represent sharing of thunks), or more complex reduction rules.

You may also want to point out that we call "reduction strategies" the more restricted reduction relations that only allow some of the possible reduction steps, and that Racket and OCaml use call-by-value.